### DISCUSSION TITLE
Alternative to permit_columns in Avram operations

### DISCUSSION THREAD

**Discussion by matthewmcgarvey** [luckyframework/lucky#1677] [Feature requests & Ideas]
Docs: https://luckyframework.org/guides/database/saving-records#allowing-params-to-be-saved-with-permit_columns

`permit_columns` defines what can come params. The caveat there is that if you defined any `attribute`'s then they can also come from params. I find this to be confusing and having watched someone try to learn Lucky, they found this confusing as well.

I propose we change the name of `permit_columns` that better represents that it is the allowed values from params and also make it to where `attribute`'s must be listed there as well or they won't be fetched.

I believe those two things will provide clarity to new users on how to define operations.

Some options for the rename
- `allowed_params`
- `allowed`
- `permit_params`
- `permitted`

**Comment by robcole** [comment]
I’ve always been fan of permitted_params as it descends from Rails, but also pretty well describes what the purpose of it is.

**Comment by jwoertink** [comment]
I agree that it can be quite confusing. Even just understanding when `permit_columns` is needed VS when it's not. The tricky part is differentiating between a model's column and the operation's attribute. When we know which is which, it makes it easier to assign the column values to the model, without trying to pass in the operation's attributes.

```crystal
# for the model
permit_columns name, email
# for this operation
attribute set_at : Time
```

Then internally we know `Model.new(**params_from_permitted_columns)` works since they're already separated.

https://github.com/luckyframework/avram/blob/c798fbbda4cd9bd40ad9800022b4c6203e6f4fe7/src/avram/save_operation.cr#L279

Now, if we were to combine them, what does that look like?

```crystal
# this doesn't work because we don't know what `set_at` is
allowed name, email, set_at

# this is probably fine, though a bit redundant since we already know what name and email are
allowed name : String, email : String, set_at : Time
```

but now how do we instantiate the model without passing it `set_at`? I guess one way we can handle that is by gathering a list of the column names from the model, and then looking at what's been allowed to create something that can be passed down. 

Another thing I just thought of is ensuring that the allowed param doesn't mismatch with the model in any way.

```crystal
table do
  column address : String
  column code : Int32
end

# I mistyped this but didn't catch it. How do we know if this is an operation attribute, or meant to be a model's column?
allowed adress : String

# or somehow not putting the type the same. Maybe I forget I had `code` in my model, and though I was adding this for the operation
allowed code : String
```

The last bit I have is that the only thing holding us back on the next release is literally params.. :sweat_smile:  This confusion is part of the reason, but also the fact that we're currently blocked from being able to explore more complex save operations due to constraints with params. (i.e. how would the "allowed" method work if you wanted to allow params for a sub object akin to rails' accepts_nested_attributes_for, etc...)

Ref: https://github.com/luckyframework/lucky/pull/1519
Ref: https://github.com/luckyframework/lucky/discussions/727
